HIP/hip-1056.md

```diff
@@ -1790,118 +1804,192 @@ A `TraceData` block item is added to contain all information that could be cosni
+ * 3. Contract slot usages, including reads and writes.
+ * 4. Events logged during the transaction.
+ * 5. Full error message, if any, produced by the contract call.
  */
-message EVMTransactionLog {
-     /**
-      * The contract emitting the log. ContractID vs 20 bye address is used to preserve space
-      */
-     ContractID contract_id = 1;
+message EvmTraceData {
+    /**
+     * The initcode executed for an internal contract creation.
+     */
+    ExecutedInitcode executed_initcode = 1;
 
-     /**
-      * The Log data
-      */
-     bytes data = 2;
+    /**
+     * All contract actions executed during the transaction.
+     */
+    repeated proto.ContractAction contract_actions = 2;
 
-     /**
-      * The logc topics left padding of 0's by EVM is stripped to save space. 
-      * Indexers should left-pad with zeros to form a 256-bit word.
-      */
-     repeated bytes topics = 3;
+    /**
+     * Contract slot usages in the transaction, by contract id.
+     */
+    repeated ContractSlotUsage contract_slot_usages = 3;
+
+    /**
+     * Log events produced during the transaction, by contract id.
+     */
+    repeated EvmTransactionLog logs = 4;
+
+    /**
+     * Additional details of any error message in the EVM transaction result.
+     * <p>
+     * This SHALL be unset if the contract call succeeded.
+     */
+    string error_details = 5;
 }
 
 /**
- * The init bytecode components for child contracts created as part of the contract execution.
- * init_bytecode = deploy_bytecode + runtime_byteode + metadata_bytecode
+ * Usage of a contract's storage slots in an EVM transaction.<br/>
+ * In the common case that a contract operation is the only EVM
+ * transaction in its Hiero transactional unit, this message simply
+ * sets `written_keys_are_non_identical_state_changes` to true;
+ * meaning a stream consumer should extract the written keys from
+ * the Hiero unit's SlotKey state changes that match this message's
+ * `contract_id` (but ignoring any* MapUpdateChange with
+ * `identical` set to true).
+ * <p>
+ * In the less common case that an EVM transaction is one of
+ * several in its Hiero transactional unit (e.g., because it is
+ * one of several ContractCalls in an AtomicBatch); this message
+ * explicitly traces the storage slots that were written during
+ * its transaction's execution.
  */
-message ContractInitByteCode {
+message ContractSlotUsage {
     /**
-     * The bytecode that predeces and deploys the runtime bytecode in a contracts deployment
+     * The contract using the storage slots.
      */
-    google.protobuf.BytesValue deploy_bytecode = 1;
+    proto.ContractID contract_id = 1;
+
+    oneof written_keys {
+        /**
+         * If true, the list of written slot keys is contained in the non-identical
+         * state changes in the Hiero transactional unit of this EVM transaction.
+         */
+        bool written_keys_are_non_identical_state_changes = 2;
```

Comment:
Do we think this is needed or could it not be inferred but the rules of the spec?
If `written_slot_keys` is not empty then slot writes are fully described here, if empty then they are in state changes. This avoids the extra bool needing to be written.
Is such a savings valuable to consider or is the explicitness required?